src/sentry/snuba/discover.py

```diff
@@ -1034,6 +1041,127 @@ def get_facets(query, params, limit=10, referrer=None):
 
     Returns Sequence[FacetResult]
     """
+    sentry_sdk.set_tag("discover.use_snql", use_snql)
+    if use_snql:
+        # temporarily add snql to referrer
+        referrer = f"{referrer}.wip-snql"
+        sample = len(params["project_id"]) > 2
+
+        key_name_builder = QueryBuilder(
+            Dataset.Discover,
+            params,
+            # Exclude tracing tags as they are noisy and generally not helpful.
+            # TODO(markus): Tracing tags are no longer written but may still reside in DB.
+            query=(query if query is not None else "")
+            + " !tags_key:[trace,trace.ctx,trace.span,project]",
+            selected_columns=["tags_key", "count()"],
+            orderby=["-count()", "tags_key"],
+            turbo=sample,
+        )
+        key_names = raw_snql_query(key_name_builder.get_snql_query(), referrer=referrer)
+        # Sampling keys for multi-project results as we don't need accuracy
+        # with that much data.
+        top_tags = [r["tags_key"] for r in key_names["data"]]
+        if not top_tags:
+            return []
+
+        sample_rate = 0.1 if (key_names["data"][0]["count"] > 10000) else None
+        # Rescale the results if we're sampling
+        multiplier = 1 / sample_rate if sample_rate is not None else 1
+
+        fetch_projects = False
+        if len(params.get("project_id", [])) > 1:
+            if len(top_tags) == limit:
+                top_tags.pop()
+            fetch_projects = True
+
+        results = []
+        if fetch_projects:
+            with sentry_sdk.start_span(op="discover.discover", description="facets.projects"):
+                project_value_builder = QueryBuilder(
+                    Dataset.Discover,
+                    params,
+                    query=query,
+                    selected_columns=["count()", "project_id"],
+                    orderby=["-count()"],
+                    turbo=sample,
```

Comment:
> turbo sets a sampling rate to the query defined in the TURBO_SAMPLE_RATE Snuba setting. It also prevents Snuba to apply the FINAL mode to the Clickhouse query in case it was needed to guarantee correct results after replacements.